### ISSUE TITLE
Stock location linked to user group

### ISSUE THREAD

**Issue by oscgonfer** [inventree/InvenTree#1129]
Hi!

At [fablab bcn](https://fablabbcn.org/) we are running an instance of [Inventree](https://inventory.fablabbcn.org/) - so far it's working great and we really appreciate the fantastic work.
Since we have various projects with different inventory locations, we wonder if there's the possibility (already implemented or not) to have a stock location linked to an user group. This facilitates the case in which an organization like ours has various projects in parallel with different teams, but you don't want them to have edit access to other stock locations, but yes view access. 

Would this be a potential enhancement? If not, we can always run different instances. 
Thanks!

**Comment by eeintech** [comment]
Hello @oscgonfer 

This is a very relevant request and is related to this issue: #656 

[django-guardian](https://github.com/django-guardian/django-guardian) seems to be a nice candidate for item ownership.

I am not quite if it can be added in the near future. Are stock items shared among your different teams?

**Comment by oscgonfer** [comment]
Hi! 

Thanks for your prompt response!

To answer the question, our structure more or less would be like this:

- robotics stock, that only robotics team uses
- sensors stock, that only sensors team uses
- electronics stock, that all the teams use
- etc

So.. it depends. Sometimes the stock items can be shared, but sometimes they are not due to budgeting and so on.

**Comment by eeintech** [comment]
I would leaned towards multiple InvenTree instances if it was a completely isolated usage of stock but in your case having similar items in different database would create a synchronization nightmare...

I don't promise anything but I really want to dig into this sooner than later, we have recently added a permission management system which definitely should be utilized to its full potential 😄

**Comment by oscgonfer** [comment]
That sounds great! I'm not great at django myself but ping me if I can help with anything

**Comment by oscgonfer** [comment]
Wow!! This is so nice! https://github.com/inventree/InvenTree/pull/1155

❤️

**Comment by eeintech** [comment]
@oscgonfer That's right! Would you have the chance to test it and let me know if that'd be fitting your requirements?
Take a look at the PR details for more information on the underlying mechanisms.

**Comment by oscgonfer** [comment]
Looks really good! 👍 

I would only say, that user groups after this are probably inherently related to stock locations, as the owner of the stock location is an user group. This makes sense in this use case, but from my point of view starts to lose the conceptual idea I perceived as an user group (more in a functional way). Either way, I think this is perfect for us and really makes our lives easier, so big big thank you for that. 

If I can contribute to anything else in testing or documentation I would be happy to help.

**Comment by eeintech** [comment]
I would **love** your eyes and time testing those new features, that'd be really great! 👍 

I decided on going for "location owner = group" and "stock item owner = user" because it was the most logical ownership implementation in my mind. Users can have multiple groups, so there is definitely doors open for use-cases like:
- group A has write access to location **X**
- group B has write access on location **Y**
- group C has write access on location **Z**
- user A is in group A and C: user A has write access to locations **X and Z** and all underlying items
- user B is in group B and C: user B has write access to locations **Y and Z** and all underlying items
- user C is in group C only: user C has write access to location **Z** and all underlying items
- item A has no location set: all users can edit it
- item B is set to location Y: only user B can edit it
- item C is set to location Z: users A, B and C can edit it

I'm open to improve the logic, if you have any suggestion 😉

**Comment by oscgonfer** [comment]
Thanks for your kind comments :)

I always think that a good documentation can explain any logic behind the code. For me it became clear after testing it in our instance, and with your example there is more than enough for understanding it if you read carefully. 

I'll set up our whole internal structure in our inventree and I'll post if I have any comments here in couple of days.

Thanks again!

**Comment by eeintech** [comment]
Wow awesome! Please do some careful backup before, just in case!

**Comment by eeintech** [comment]
@oscgonfer Just wanted to give you an update as why this isn't merged into master yet... sorry :disappointed: 
@SchrodingersGat made some good proposals and I summarized them in the PR #1155 in a "story-telling" table which hopefully covers all ownership use-cases.

I have had little time in my hands for coding those last few weeks and won't probably be able to go forward with the improvements until next year. What I can hope is that implementing the changes won't affect the core architecture of this feature, so I would think you should be okay to go forward and use my branch for the time being (I will try to keep it up-to-date with master).

**Comment by oscgonfer** [comment]
Hi! No worries at all, I have been keeping an eye on your updates and merging locally on my branch. I really appreciate all the work you are putting on this. I had a look at your summary and now has grown further than what I was initially expecting, but that's a great thing. I'll keep an eye and a frequent backup just in case. Thanks!

**Comment by eeintech** [comment]
@oscgonfer Alright I think I have something ready in #1155

If you have the chance, pull it and test it as much as possible, any feedback would be greatly appreciated :smiley:

**Comment by oscgonfer** [comment]
Hi @eeintech!

Thanks so much for this. Pulling now and testing it with the team this week. I will put comments here if any.

**Comment by oscgonfer** [comment]
Hey, 

just tried an update following these steps and came across a problem:

```
sudo systemctl stop gunicorn.socket
git pull origin pull/1155/head
```

When running:

```
inv update
```

The following issue came up:

```
[...]
2021-01-17 10:36:29,155 INFO InvenTree running in DEBUG mode
2021-01-17 10:36:29,155 INFO MEDIA_ROOT: '/home/inventree/inventree_media'
2021-01-17 10:36:29,155 INFO STATIC_ROOT: '/home/inventree/inventree_static'
2021-01-17 10:36:29,155 INFO Configuring database backend:
2021-01-17 10:36:29,155 INFO Database ENGINE: 'django.db.backends.postgresql_psycopg2'
2021-01-17 10:36:29,155 INFO Database NAME: 'Inventree_db'
Checking Company image thumbnails
Using selector: EpollSelector
InvenTree: Checking Part image thumbnails
Loading action plugins...
Loading action plugins
Discovered 1 action plugins:
 - SimpleActionPlugin
Operations to perform:
  Apply all migrations: admin, auth, authtoken, build, common, company, contenttypes, error_report, exchange, label, order, part, report, sessions, stock, users
Running migrations:
  Applying stock.0057_stock_location_item_owner...Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/django/db/backends/utils.py", line 86, in _execute
    return self.cursor.execute(sql, params)
psycopg2.errors.DuplicateColumn: column "owner_id" of relation "stock_stockitem" already exists


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "manage.py", line 22, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.8/dist-packages/django/core/management/__init__.py", line 401, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.8/dist-packages/django/core/management/__init__.py", line 395, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.8/dist-packages/django/core/management/base.py", line 328, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.8/dist-packages/django/core/management/base.py", line 369, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/python3.8/dist-packages/django/core/management/base.py", line 83, in wrapped
    res = handle_func(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/django/core/management/commands/migrate.py", line 231, in handle
    post_migrate_state = executor.migrate(
  File "/usr/local/lib/python3.8/dist-packages/django/db/migrations/executor.py", line 117, in migrate
    state = self._migrate_all_forwards(state, plan, full_plan, fake=fake, fake_initial=fake_initial)
  File "/usr/local/lib/python3.8/dist-packages/django/db/migrations/executor.py", line 147, in _migrate_all_forwards
    state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
  File "/usr/local/lib/python3.8/dist-packages/django/db/migrations/executor.py", line 245, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/usr/local/lib/python3.8/dist-packages/django/db/migrations/migration.py", line 124, in apply
    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
  File "/usr/local/lib/python3.8/dist-packages/django/db/migrations/operations/fields.py", line 110, in database_forwards
    schema_editor.add_field(
  File "/usr/local/lib/python3.8/dist-packages/django/db/backends/base/schema.py", line 480, in add_field
    self.execute(sql, params)
  File "/usr/local/lib/python3.8/dist-packages/django/db/backends/base/schema.py", line 142, in execute
    cursor.execute(sql, params)
  File "/usr/local/lib/python3.8/dist-packages/django/db/backends/utils.py", line 100, in execute
    return super().execute(sql, params)
  File "/usr/local/lib/python3.8/dist-packages/django/db/backends/utils.py", line 68, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/usr/local/lib/python3.8/dist-packages/django/db/backends/utils.py", line 77, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/usr/local/lib/python3.8/dist-packages/django/db/backends/utils.py", line 86, in _execute
    return self.cursor.execute(sql, params)
  File "/usr/local/lib/python3.8/dist-packages/django/db/utils.py", line 90, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/usr/local/lib/python3.8/dist-packages/django/db/backends/utils.py", line 86, in _execute
    return self.cursor.execute(sql, params)
django.db.utils.ProgrammingError: column "owner_id" of relation "stock_stockitem" already exists
```

**Comment by eeintech** [comment]
@oscgonfer Did you migrate/update your database with a previous version of the PR? This error is most likely because of that, and the fact that the owner field was different at that time.
I think you should delete the `owner` column  in the `stock_stockitem` table, it's very likely you'll have to do this with the `stock_stocklocation` table too.

**Comment by oscgonfer** [comment]
Hi!

Yes, I performed `inv update` after the `git pull` of your PR. 
Regarding the table columns, is it `owner_id` that I have to delete? I didn't see `owner` column, unless I misunderstood:

I performed `\d+ stock_stockitem;` and saw that the columns are: 

```
id                      
serial                  
link                    
batch                   
quantity                
updated                 
stocktake_date          
review_needed           
delete_on_deplete       
status                  
notes                   
infinite                
belongs_to_id           
location_id             
part_id                 
stocktake_user_id       
supplier_part_id        
purchase_order_id       
build_id                
level                   
lft                     
parent_id               
rght                    
tree_id                 
uid                     
sales_order_id          
customer_id             
is_building             
purchase_price          
purchase_price_currency 
owner_id                
expiry_date
```

And similar to `stock_stocklocation`:

```
id
name
description
parent_id
level
lft
rght
tree_id
owner_id
```

**Comment by eeintech** [comment]
@oscgonfer Yes, it's the `owner_id` column, sorry. Please delete it and re-run `inv migrate`.
As you already had this column in your database it means that it was not the first time you pulled the PR code and migrated your database.

`inv update` does `install` (update dependencies), `migrate` (database update) and `static` (collect static files) for you in one command. To see if the problem is fixed, after deleting the column, you only need `migrate`.
